### PR TITLE
add lifecycle datasource

### PR DESCRIPTION
**Pull request by taimoorgit**
## Issues

https://github.com/OpsLevel/team-platform/issues/288

## Tophatting

```
data "opslevel_lifecycle" "l1" {
  filter {
    field = "alias"
    value = "generally_available"
  }
}

data "opslevel_lifecycle" "l2" {
  filter {
    field = "id"
    value = "Z2lkOi8vb3BzbGV2ZWwvTGlmZWN5Y2xlLzQ"
  }
}

output "o1" {
  value = data.opslevel_lifecycle.l1
}

output "o2" {
  value = data.opslevel_lifecycle.l2
}
```

```
Changes to Outputs:
  + o1 = {
      + alias       = "generally_available"
      + description = "Service is supporting features available to be used by all customers, and should be fully stable."
      + filter      = {
          + field = "alias"
          + value = "generally_available"
        }
      + id          = "Z2lkOi8vb3BzbGV2ZWwvTGlmZWN5Y2xlLzQ"
      + index       = 4
      + name        = "Generally Available"
    }
  + o2 = {
      + alias       = "generally_available"
      + description = "Service is supporting features available to be used by all customers, and should be fully stable."
      + filter      = {
          + field = "id"
          + value = "Z2lkOi8vb3BzbGV2ZWwvTGlmZWN5Y2xlLzQ"
        }
      + id          = "Z2lkOi8vb3BzbGV2ZWwvTGlmZWN5Y2xlLzQ"
      + index       = 4
      + name        = "Generally Available"
    }

```

### CODE REVIEW

**Comment by taimoorgit** [comment]
Index filter also works:

```
data "opslevel_lifecycle" "l3" {
  filter {
    field = "index"
    value = "4"
  }
}

```

```
  + o3 = {
      + alias       = "generally_available"
      + description = "Service is supporting features available to be used by all customers, and should be fully stable."
      + filter      = {
          + field = "id"
          + value = "Z2lkOi8vb3BzbGV2ZWwvTGlmZWN5Y2xlLzQ"
        }
      + id          = "Z2lkOi8vb3BzbGV2ZWwvTGlmZWN5Y2xlLzQ"
      + index       = 4
      + name        = "Generally Available"
    }

```

File: tests/mock_datasource/lifecycle.tfmock.hcl [hunk 1]
```diff
@@ -0,0 +1,7 @@
+mock_data "opslevel_lifecycle" {
+  defaults = {
```
Comment (davidbloss) [inline tests/mock_datasource/lifecycle.tfmock.hcl:2]:
```suggestion
  defaults = {
    alias = "Lifecycle Alias"
```

File: opslevel/datasource_opslevel_lifecycle.go [hunk 2]
```diff
@@ -1,94 +1,141 @@
...
+				Computed:    true,
+			},
+		},
+		Blocks: map[string]schema.Block{
+			"filter": getDatasourceFilter(validFieldNames),
+		},
+	}
+}
+
+func (sys *LifecycleDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
```
Comment (davidbloss) [inline opslevel/datasource_opslevel_lifecycle.go:86]: Spotted some copy pasta 🍝  `sys` maybe could be `d` or something

File: opslevel/datasource_opslevel_lifecycle.go [hunk 3]
```diff
@@ -1,94 +1,141 @@
...
+	}
+
+	lifecycleDataModel := NewLifecycleDataSourceModel(ctx, *lifecycle, data.Filter)
+
+	// Save data into Terraform state
+	tflog.Trace(ctx, "read an OpsLevel Lifecycle data source")
+	resp.Diagnostics.Append(resp.State.Set(ctx, &lifecycleDataModel)...)
+}
+
+func filterLifecycles(tiers []opslevel.Lifecycle, filter FilterBlockModel) (*opslevel.Lifecycle, error) {
```
Comment (davidbloss) [inline opslevel/datasource_opslevel_lifecycle.go:114]: Moar copy pasta 🍝  `tiers` --> `lifecycles`

**Review by davidbloss — CHANGES_REQUESTED** [review]
Overall looks good! Just needs a few tweaks

File: tests/datasource_lifecycle.tftest.hcl [hunk 4]
```diff
@@ -8,6 +8,11 @@ run "datasource_lifecycle_filter_by_id" {
     opslevel = opslevel.fake
   }
 
+  assert {
+    condition     = data.opslevel_lifecycle.id_filter.alias == "generally_available"
+    error_message = "id in opslevel_lifecycle mock was not set"
```
Comment (davidbloss) [inline tests/datasource_lifecycle.tftest.hcl:13]:
```suggestion
    error_message = "wrong alias in opslevel_lifecycle mock"
```

File: tests/datasource_lifecycle.tftest.hcl [hunk 5]
```diff
@@ -35,6 +40,11 @@ run "datasource_lifecycle_filter_by_name" {
     opslevel = opslevel.fake
   }
 
+  assert {
+    condition     = data.opslevel_lifecycle.name_filter.alias == "generally_available"
+    error_message = "id in opslevel_lifecycle mock was not set"
```
Comment (davidbloss) [inline tests/datasource_lifecycle.tftest.hcl:45]:
```suggestion
    error_message = "wrong alias in opslevel_lifecycle mock"
```

File: tests/datasource_lifecycle.tftest.hcl [hunk 6]
```diff
@@ -52,6 +62,11 @@ run "datasource_lifecycle_filter_by_index" {
     opslevel = opslevel.fake
   }
 
+  assert {
+    condition     = data.opslevel_lifecycle.index_filter.alias == "generally_available"
+    error_message = "id in opslevel_lifecycle mock was not set"
```
Comment (davidbloss) [inline tests/datasource_lifecycle.tftest.hcl:67]:
```suggestion
    error_message = "wrong alias in opslevel_lifecycle mock"
```

File: tests/datasource_lifecycle.tftest.hcl [hunk 7]
```diff
@@ -69,6 +84,11 @@ run "datasource_lifecycle_filter_by_alias" {
     opslevel = opslevel.fake
   }
 
+  assert {
+    condition     = data.opslevel_lifecycle.index_filter.alias == "generally_available"
```
Comment (davidbloss) [inline tests/datasource_lifecycle.tftest.hcl:88]:
```suggestion
    error_message = "wrong alias in opslevel_lifecycle mock"
```

File: tests/datasource_lifecycle.tftest.hcl [hunk 8]
```diff
@@ -69,6 +84,11 @@ run "datasource_lifecycle_filter_by_alias" {
...
+    error_message = "id in opslevel_lifecycle mock was not set"
```
[the remark this answers is not in the document]
Comment (davidbloss) [reply tests/datasource_lifecycle.tftest.hcl:89]: Under the hood, these four `assert { }` blocks are all testing the same `.alias` field defined in `tests/mock_datasource/lifecycle.tfmock.hcl`

**Comment by davidbloss** [comment]
Latest update looks better! I should have check the first time but there is no `description` field in our current [lifecycle datasource](https://registry.terraform.io/providers/OpsLevel/opslevel/latest/docs/data-sources/lifecycle), or in the old code

**Comment by taimoorgit** [comment]
> Latest update looks better! I should have check the first time but there is no `description` field in our current [lifecycle datasource](https://registry.terraform.io/providers/OpsLevel/opslevel/latest/docs/data-sources/lifecycle), or in the old code

Just pushed a change to remove description

**Review by davidbloss — APPROVED** [review]
LGTM! 🎉

**Pull request merged**